pkg/replica/hash.go

```diff
@@ -0,0 +1,340 @@
+
+	lastHashedAt := time.Now().UTC().Format(time.RFC3339)
+	checksum, err = hashSnapshot(t.Ctx, t.SnapshotName)
+	if err != nil {
+		return
+	}
+
+	if t.isSnapshotSilentlyCorrupted(checksum) {
+		t.StatusLock.Lock()
+		t.SilentlyCorrupted = true
+		t.StatusLock.Unlock()
+		return
+	}
+
+	logrus.Infof("Snapshot %v checksum %v", t.SnapshotName, checksum)
+
+	err = SetSnapshotHashInfoToXattr(t.SnapshotName, checksum, modTime, lastHashedAt)
+	if err != nil {
+		return
+	}
+
+	err = t.isModTimeRemain(modTime)
+	if err != nil {
+		err = DeleteSnapshotHashInfoFromXattr(t.SnapshotName)
+	}
+}
+
+func (t *SnapshotHashTask) isSnapshotSilentlyCorrupted(checksum string) bool {
+	// To detect the silent corruption, read the modTime and checksum already recorded in the snapshot disk file first.
+	// Then, rehash the file and compare the modTimes and checksums.
+	// If the modTimes are identical but the checksums differ, the file is silently corrupted.
+
+	existingChecksum := ""
+	existingModTime := ""
+
+	existingChecksum, existingModTime, err := GetSnapshotHashInfoFromXattr(t.SnapshotName)
+	if err != nil && err != syscall.ENODATA {
+		return false
+	}
+
+	if existingChecksum == "" || existingModTime == "" {
+		return false
+	}
+
+	if err := t.isModTimeRemain(existingModTime); err != nil {
+		return false
+	}
+
+	if checksum != existingChecksum {
+		return true
+	}
+
+	return false
+}
+
+func GetSnapshotModTime(snapshotName string) (string, error) {
+	fileInfo, err := os.Stat(diskutil.GenerateSnapshotDiskName(snapshotName))
+	if err != nil {
+		return "", err
+	}
+
+	return fileInfo.ModTime().String(), nil
+}
+
+func GetSnapshotHashInfoFromXattr(snapshotName string) (string, string, error) {
+	xattrSnapshotHashValue := make([]byte, xattrSnapshotHashValueMax)
+	_, err := unix.Getxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue)
+	if err != nil {
+		return "", "", err
+	}
+
+	index := bytes.IndexByte(xattrSnapshotHashValue, 0)
+
+	info := &SnapshotXattrHashInfo{}
+	if err := json.Unmarshal(xattrSnapshotHashValue[:index], info); err != nil {
+		return "", "", err
+	}
+
+	return info.Checksum, info.ModTime, nil
+}
+
+func SetSnapshotHashInfoToXattr(snapshotName, checksum, modTime, lastHashedAt string) error {
+	xattrSnapshotHashValue, err := json.Marshal(&SnapshotXattrHashInfo{
+		Method:       defaultHashMethod,
+		Checksum:     checksum,
+		ModTime:      modTime,
+		LastHashedAt: lastHashedAt,
+	})
+	if err != nil {
+		return err
+	}
+
+	return unix.Setxattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName, xattrSnapshotHashValue, 0)
+}
+
+func DeleteSnapshotHashInfoFromXattr(snapshotName string) error {
+	return unix.Removexattr(diskutil.GenerateSnapshotDiskName(snapshotName), xattrSnapshotHashName)
+}
+
+func (t *SnapshotHashTask) getSize() (int64, error) {
+	fileInfo, err := t.file.Stat()
+	if err != nil {
+		return -1, err
+	}
+
+	return fileInfo.Size(), nil
+}
```

Comment:
What's this for? BTW, this returns the apparent file size.

Reply:
Dead codes. Will remove it.